Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+```scala
+// here both query() and map() calls accept the execution context implicitly
+def load(id: String)(implicit ec: ExecutionContext): Future[SomeObject] =
+  remoteApi.query(id).map(convertFromJson)
+```
+
+However, it is almost never necessary to use implicit parameters of the types defined in the custom code base. Oftentimes this is done to "simplify" the code which eventually results in code which is messy and hard to read. Therefore, using implicit parameters for custom types unrelated to other libraries should be avoided.
+
+When interacting with the third-party libraries, it is sometimes necessary to define new implicit objects. The most prominent example is defining serializers for custom types. Definition of these implicits must be as contained as possible. Ideally a set of such definitions which are related in some way should be grouped together in one object, and this object should be imported when these implicits are needed:
+
+```scala
+object ModelJsonFormatters {
+  implicit val userFormat = Json.format[User]
+  implicit val serviceFormat = Json.format[Service]
+  implicit val whateverFormat = Json.format[Whatever]
+}
+
+// at the use site
+import some.pkg.ModelJsonFormatters._
+
+val jsonData = Json.toJson(Whatever(...))
+```
+
+Naturally, using them must not contradict the rules about imports as they are defined above.
+
+As a final note, implicits are sometimes feared because they are used in complex libraries like scalaz. However, these libraries actually provide *concepts* which are hard to understand without prior experience; implicits only allow these libraries not to look absolutely horrible, and their absence from there would not make things better. This is the reason why usage of implicits is allowed: if used correctly (according to the guidelines presented above), they lead to much clearer code. If they are used together with a library which itself is very complex, they make things worse. The solution is to avoid using such libraries but not implicits; more on this below.
+
+## `override` modifier
+
+Scala requires `override` modifiers on items which override items with the same defined in the parent class, but it does not require it if these parent items are `abstract`:
+
+```scala
+abstract class Root {
+  def x: Int = 10
+  abstract def y: Int
+}
+
+class Child extends Root {
+  override def x: Int = 12
+  def y: Int = 42
+}
+```
+
+However, you should always use `override` modifier even for implementing abstract methods. This makes the fact that the item actually "comes" from a parent class obvious and it allows accidental non-overrides, for example, as a side effect of refactoring, to be caught during compilation easily.
+
+Note that you must use the `override` modifier even if you define the overriding field as a part of constructor arguments declaration:
+
+```scala
+trait Entity {
+  def id: String
+}
+
+case class User(override val id: String) extends Entity
```

Comment:
In case when you have to declare multiple fields which are overriding items from the trait, it would be quite natural to group them together, preferably at the beginning of the parameters list:
```scala
case class FooBar(override val name: String,
                  override val somethingElse: OtherType,
                  override val isBad: Boolean,
                  id: String,
                  age: Int,
                  isGood: Boolean) extends WhatEver
```
And I believe when it is written like this, the readability does not suffer *that* much to compensate for the absence of `override`.